app/models/credential_factory.rb

```diff
@@ -0,0 +1,49 @@
+module CredentialFactory
+  class << self
+    def values factory_resource
+      provider_annotation = factory_resource.credential_factory_provider
+      raise ArgumentError, %Q(Annotation "credential-factory/provider" not found on #{factory_resource.id.inspect}) unless provider_annotation
+      require "credential_factory/#{provider_annotation.value.underscore}"
+      factory_class = self.const_get(provider_annotation.value.downcase.camelize)
+
+      annotations = factory_resource.annotations.select do |annotation|
+        annotation.name =~ /^credential-factory\// && annotation.name != "credential-factory/provider"
+      end.inject({}) do |memo, annotation|
+        memo[annotation.name] = annotation.value
+        memo
+      end
+
+      dependent_variable_ids = factory_class.dependent_variable_ids(annotations)
+
+      dependent_variables = dependent_variable_ids.map do |id|
+        tokens = id.split(":")
+        if tokens.length == 1
+          id = [ factory_resource.account, "variable", id ].join(":")
+        end
+
+        Resource[id].tap do |variable|
+          unless variable
+            Rails.logger.info "Variable #{id.inspect} required by #{factory_resource.resource_id.inspect} does not exist"
+            raise Exceptions::RecordNotFound, id
+          end
+        end
+      end
+      dependent_secrets = dependent_variables.inject({}) do |memo, variable|
```

Comment:
Protip: instead of
```ruby
xh = xs.inject({}) do |memo, x|
  memo[k] = f(x)
  memo
end
```
you can do
```ruby
xh = xs.map do |x|
  [k, f(x)]
end.to_h
```